lib/p5.play.js

```diff
@@ -2189,246 +2123,230 @@ function Sprite(pInst, _x, _y, _w, _h) {
+   * @param {Sprite} target
+   * @param {function} callback - if collision occurred
+   * @return {boolean} true if a collision occurred
+   */
+  this._collideWithSprite = function(type, target, callback) {
+    // Never collide with self
+    if (target === this) {
+      return false;
+    }
 
-              //the extents are the distance between the coll centers
-              //plus the extents of both
-              var e = createVector(
-                abs(this.position.x -this.previousPosition.x) + this.collider.extents.x,
-                abs(this.position.y -this.previousPosition.y) + this.collider.extents.y);
+    if (this.collider === undefined) {
+      this.setDefaultCollider();
+    }
 
-              var bbox = new AABB(pInst, c, e, this.collider.offset);
+    if (target.collider === undefined) {
+      target.setDefaultCollider();
+    }
 
-              //bbox.draw();
+    if (!this.collider || !target.collider) {
+      // We were unable to create a collider for one of the sprites.
+      // This usually means its animation is not available yet; it will be soon.
+      // Don't collide for now.
+      return false;
+    }
 
-              if(bbox.overlap(other.collider))
-              {
-                if(tunnelX) {
+    // Actually compute the overlap of the two colliders
+    var displacement = this._findDisplacement(target);
+    if (displacement.x === 0 && displacement.y === 0) {
+      // These sprites are not overlapping.
+      return false;
+    }
 
-                  //entering from the right
-                  if(this.velocity.x < 0)
-                    displacement.x = other.collider.right() - this.collider.left() + 1;
-                  else if(this.velocity.x > 0 )
-                    displacement.x = other.collider.left() - this.collider.right() -1;
-                  }
+    if (type === 'collide' || type === 'displace' || type === 'bounce') {
+      if (type === 'displace' && !target.immovable) {
+        target.position.sub(displacement);
+      } else if ((type === 'collide' || type === 'bounce') && !this.immovable) {
+        this.position.add(displacement);
+        this.previousPosition = createVector(this.position.x, this.position.y);
+        this.newPosition = createVector(this.position.x, this.position.y);
+        this.collider.updateFromSprite(this);
+      }
 
-                if(tunnelY) {
-                  //from top
-                  if(this.velocity.y > 0)
-                    displacement.y = other.collider.top() - this.collider.bottom() - 1;
-                  else if(this.velocity.y < 0 )
-                    displacement.y = other.collider.bottom() - this.collider.top() + 1;
+      if (displacement.x > 0)
+        this.touching.left = true;
+      if (displacement.x < 0)
+        this.touching.right = true;
+      if (displacement.y < 0)
+        this.touching.bottom = true;
+      if (displacement.y > 0)
+        this.touching.top = true;
+
+      if (type === 'bounce') {
```

Comment:
I think the problem here is that the newVel1 and newVel2 formulae are a little too simple - see https://en.wikipedia.org/wiki/Inelastic_collision#Formula, it looks like you've got the first term but you're missing the second two.